File: 剑指offer/41.数据流中的中位数.py
import heapq


class MedianFinder:
    def __init__(self):
        """
        initialize your data structure here.
        """
        # 小顶堆存储较大的数
        self.lager = []
        # 大顶堆存储较小的数
        self.smaller = []

    def addNum(self, num: int) -> None:
        if len(self.lager) != len(self.smaller):
            heapq.heappush(self.lager, num)
            heapq.heappush(self.smaller, heapq.heappop(self.lager))
        else:
            heapq.heappush(self.smaller, num)
            heapq.heappush(self.lager, heapq.heappop(self.smaller))

    def findMedian(self) -> float:
        if len(self.lager) != len(self.smaller):
            return self.lager[0]
        else:
            return (self.lager[0] + self.smaller[0]) // 2


# Your MedianFinder object will be instantiated and called as such:
# obj = MedianFinder()
# obj.addNum(num)
# param_2 = obj.findMedian()

File: 剑指offer/41.数据流中的中位数.cpp
#include <queue>
using namespace std;

class MedianFinder {
public:
    priority_queue<int, vector<int>, less<int>> maxHeap; // 大顶堆，存储较小的一半
    priority_queue<int, vector<int>, greater<int>> minHeap; // 小顶堆，存储较大的一半
    /** initialize your data structure here. */
    MedianFinder()
    {
    }

    void addNum(int num)
    {
        // 插入到较大的一边
        if (maxHeap.size() == minHeap.size()) {
            maxHeap.push(num);
            minHeap.push(maxHeap.top());
            maxHeap.pop();
        } else {
            minHeap.push(num);
            maxHeap.push(minHeap.top());
            minHeap.pop();
        }
    }

    double findMedian()
    {
        if (maxHeap.size() == minHeap.size()) {
            return (minHeap.top() + maxHeap.top()) / 2.0;
        } else {
            return minHeap.top() * 1.0;
        }
    }
};
